runtime/src/axum/mod.rs

```diff
@@ -0,0 +1,529 @@
+use shuttle_proto::runtime::runtime_server::Runtime;
+use shuttle_proto::runtime::{
+    self, LoadRequest, LoadResponse, StartRequest, StartResponse, StopRequest, StopResponse,
+    SubscribeLogsRequest,
+};
+use shuttle_service::ServiceName;
+use tokio::sync::mpsc::{Receiver, Sender};
+use tokio::sync::{mpsc, oneshot};
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::Status;
+use tracing::{error, trace};
+use wasi_common::file::FileCaps;
+use wasmtime::{Engine, Linker, Module, Store};
+use wasmtime_wasi::sync::net::UnixStream as WasiUnixStream;
+use wasmtime_wasi::{WasiCtx, WasiCtxBuilder};
+
+extern crate rmp_serde as rmps;
+
+const LOGS_FD: u32 = 20;
+const PARTS_FD: u32 = 3;
+const BODY_WRITE_FD: u32 = 4;
+const BODY_READ_FD: u32 = 5;
+
+pub struct AxumWasm {
+    router: Mutex<Option<Router>>,
+    logs_rx: Mutex<Option<Receiver<Result<runtime::LogItem, Status>>>>,
+    logs_tx: Mutex<Sender<Result<runtime::LogItem, Status>>>,
+    kill_tx: Mutex<Option<oneshot::Sender<String>>>,
+}
+
+impl AxumWasm {
+    pub fn new() -> Self {
+        // Allow about 2^15 = 32k logs of backpressure
+        // We know the wasm currently handles about 16k requests per second (req / sec) so 16k seems to be a safe number
+        // As we make performance gains elsewhere this might eventually become the new bottleneck to increase :D
+        //
+        // Testing has shown that a number half the req / sec yields poor performance. A number the same as the req / sec
+        // seems acceptable so going with double the number for some headroom
+        let (tx, rx) = mpsc::channel(1 << 15);
+
+        Self {
+            router: Mutex::new(None),
+            logs_rx: Mutex::new(Some(rx)),
+            logs_tx: Mutex::new(tx),
+            kill_tx: Mutex::new(None),
+        }
+    }
+}
+
+impl Default for AxumWasm {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[async_trait]
+impl Runtime for AxumWasm {
+    async fn load(
+        &self,
+        request: tonic::Request<LoadRequest>,
+    ) -> Result<tonic::Response<LoadResponse>, Status> {
+        let wasm_path = request.into_inner().path;
+        trace!(wasm_path, "loading");
+
+        let router = RouterBuilder::new()
+            .map_err(|err| Status::from_error(err.into()))?
+            .src(wasm_path)
+            .build()
+            .map_err(|err| Status::from_error(err.into()))?;
+
+        *self.router.lock().unwrap() = Some(router);
+
+        let message = LoadResponse { success: true };
+
+        Ok(tonic::Response::new(message))
+    }
+
+    async fn start(
+        &self,
+        request: tonic::Request<StartRequest>,
+    ) -> Result<tonic::Response<StartResponse>, Status> {
+        let StartRequest {
+            deployment_id,
+            port,
+            ..
+        } = request.into_inner();
+
+        let address = SocketAddr::new(Ipv4Addr::LOCALHOST.into(), port as u16);
+
+        let logs_tx = self.logs_tx.lock().unwrap().clone();
```

Comment:
Comment missing for why `unwrap()` is safe to call

Reply:
I might be mistaken but I think there is some consensus on `unwrap` of `.lock` being okay, since it will only panic if something else panicked. We could also expect I suppose.